src/docfx/lib/git/RepositoryProvider.cs

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+
+namespace Microsoft.Docs.Build
+{
+    internal class RepositoryProvider
+    {
+        private readonly ConcurrentDictionary<string, Repository> _repositoryByFolder = new ConcurrentDictionary<string, Repository>();
+
+        private readonly ConcurrentDictionary<string, GitCommitProvider> _gitCommitProvider = new ConcurrentDictionary<string, GitCommitProvider>();
+
+        public GitCommitProvider GetGitCommitProvider(string repoPath, string repoRemote = null, ConcurrentDictionary<string, Dictionary<(long commit, long blob), (long[] commitHistory, int lruOrder)>> commitCache = null)
```

Comment:
if I first call this method without a commit cache and then with a commit cache using the same repo, the second call will give back the an instance without cache, that is incorrect. Same is true if `repoRemote` is passed differently for the same repo.

Reply:
We could share a global GitCommitProvider (always with cache) and expose it like:

```csharp
class GitCommitProvider
{
    private readonly Lazy<Task<Dictionary>> _commitCache;

    public GitCommitProvider(string repoPath);

    // This one uses cache and returns full history
    public Task<List<GitCommit>> GetCommitHistory(string file, string committish = null);

    // This one does not use cache and return only the TOP N history
    public List<GitCommit> GetCommitHistory(string file, int n, string committish = null);
}
```